File: spectral/sparse_modeling/arXiv-2004-02426v2/anc/SpM/julia/input.jl
#=
!***********************************************************
! Sparse modeling approach to analytic continuation
!      of imaginary-time Monte Carlo data
!                            3 Apr 2020  E.Itou and Y. Nagai
!***********************************************************
=#

module Input
    struct Inputdata
        filename::String
        data

        function Inputdata(filename)
            data = readlines(filename)
            return new(filename,data)
        end
    end
    export Inputdata,getFloat64,getInt,getString

    function getvalue(input::Inputdata,type,name::String)
        for i=1:length(input.data)
            u = split(input.data[i])
            len = length(u[name .== u])
            if len != 0
                return parse(type,u[3])
            end
        end        
        return nothing
    end

    function getInt(input::Inputdata,name::String)
        return getvalue(input,Int64,name)
    end

    function getFloat64(input::Inputdata,name::String,initialvalue)
        v = getvalue(input,Float64,name)
        if v === nothing
            return initialvalue
        end
        return v
    end

    function getInt(input::Inputdata,name::String,initialvalue)
        n = getvalue(input,Int64,name)
        if n === nothing
            return initialvalue
        end
        return n
    end

    function getString(input::Inputdata,name::String,initialvalue)
        for i=1:length(input.data)
            u = split(input.data[i])
            len = length(u[name .== u])
            if len != 0
                return u[3]
            end
        end        
        return initialvalue
    end

end

File: spectral/sparse_modeling/arXiv-2004-02426v2/anc/SpM/c++/input.cpp
//***********************************************************
// Sparse modeling approach to analytic continuation
//      of imaginary-time Monte Carlo data
//                           3 Apr 2020  E. Itou and Y. Nagai
//***********************************************************

#include <iostream>
#include <fstream>
#include <sstream>

#include "input.h"


Input::Input(int argc, char* argv[]){
    using namespace std;
    if(argc != 2){
        cerr << "num of parameters is wrong." << endl;
        cerr << "./sparse_c input.in" << endl;
    }

    _filename = argv[1];
    _ifs.open(_filename, std::ios::in);
    std::cout << _filename << std::endl; 
    if(!_ifs){
        std::cerr << "Error: file not opened." << std::endl;
    }

};

int Input::getInt(std::string name,int initialvalue){
    std::string tmp;
    std::string strname;
    std::string strtmp;
    int num;
    

    while(getline(_ifs, tmp)){
        //std::cout << tmp << std::endl; 
        std::stringstream ss;
        ss << tmp;
        ss >> strname >> strtmp >> num; 
        if(strname == name){
            _ifs.close();
            _ifs.open(_filename, std::ios::in);
            return num;
//            std::cout << strname << "nameok" << std::endl; 
        }
        
        //std::cout << num << std::endl; 
    }
    _ifs.close();
    _ifs.open(_filename, std::ios::in);
    //_ifs.close();

    return initialvalue;
}

std::string Input::getstring(std::string name,std::string initialvalue){
    std::string tmp;
    std::string strname;
    std::string strtmp;
    std::string st;

    while(getline(_ifs, tmp)){
        std::cout << tmp << std::endl; 
        std::stringstream ss;
        ss << tmp;
        ss >> strname >> strtmp >> st; 
        if(strname == name){
            _ifs.close();
            _ifs.open(_filename, std::ios::in);
            return st;
//            std::cout << strname << "nameok" << std::endl; 
        }
        
        //std::cout << st << std::endl; 
    }
    _ifs.close();
    _ifs.open(_filename, std::ios::in);
    

    return initialvalue;
}

bool checkFileExistence(const std::string& str)
{
    std::ifstream ifs(str);
    return ifs.is_open();
}

